### PR TITLE
Definition from application.properties to Java field which have Config* annotation

### LINKED ISSUE #4: Definition from application.properties to Java field which have Config* annotation
Jump to the definition from application.properties to Java field which have Config* annotation, only when source is available.

### PR DESCRIPTION
**Pull request by angelozerr**
Definition from application.properties to Java field which have Config* annotation

Fixes #4

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR requires the vscode-qarkus https://github.com/redhat-developer/vscode-quarkus/pull/98

**Comment by angelozerr** [comment]
Here a demo:

QuarkusDefinitionDemo

**Comment by angelozerr** [comment]
@fbricon all tests for definition has been writen:

 * on JDT LS side, I test definition for JAR, deployment JAR and Java sources. See https://github.com/redhat-developer/quarkus-ls/pull/104/files#diff-804575c31ec123585056c9ea7911508fR45

 * on Quarkus LS side, I test Quarkus language service definition on comments, key and value. See https://github.com/redhat-developer/quarkus-ls/pull/104/files#diff-f3848e8d627cf2c0e734c46894e71307R31

This last test uses the JSON https://github.com/redhat-developer/quarkus-ls/pull/104/files#diff-b71ddfb095443d46349a5aa2958cb0d6 which contains all location of all Quarkus JAR which have jdt uri.

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ */
+public class MavenArtifactResolver implements ArtifactResolver {
+
+	@Override
+	public String getArtifact(String groupId, String artifactId, String version) {
+		File jarFile = null;
+		try {
+			jarFile = DependencyUtil.getArtifact(groupId, artifactId, version, null);
+		} catch (Exception e) {
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java]: exception should be logged

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+		}
+		return jarFile != null ? jarFile.toString() : null;
+	}
+
+	@Override
+	public String getSources(String groupId, String artifactId, String version) {
+		File jarFile = null;
+		try {
+			jarFile = DependencyUtil.getSources(groupId, artifactId, version);
+		} catch (Exception e) {
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java]: exception should be logged

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/QuarkusDeploymentJavaProject.java [hunk 3]
```diff
@@ -42,26 +41,19 @@
...
-	public static final ArtifactResolver MAVEN_ARTIFACT_RESOLVER = (groupId, artifactId, version) -> {
-		File jarFile = null;
-		try {
-			jarFile = DependencyUtil.getArtifact(groupId, artifactId, version, null);
-		} catch (Exception e) {
-			return null;
-		}
-		return jarFile != null ? jarFile.toString() : null;
-	};
+	public static final ArtifactResolver MAVEN_ARTIFACT_RESOLVER = new MavenArtifactResolver();
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/QuarkusDeploymentJavaProject.java]: public static final ArtifactResolver DEFAULT

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/ls/QuarkusDelegateCommandHandler.java [hunk 4]
```diff
@@ -118,4 +130,50 @@ private static DocumentationConverter getDocumentationConverter(String[] documen
 		return DocumentationConverter.DEFAULT_CONVERTER;
 	}
 
+	private static Location findDeclaredQuarkusProperty(List<Object> arguments, String commandId,
+			IProgressMonitor progress) throws CoreException {
+		Map<String, Object> obj = arguments.size() > 0 ? (Map<String, Object>) arguments.get(0) : null;
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/ls/QuarkusDelegateCommandHandler.java]: arguments.isEmpty()

File: quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/fruits.html [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<!doctype html>
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/fruits.html]: file not necessary to compile

File: quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/docker/Dockerfile.jvm [hunk 6]
```diff
@@ -0,0 +1,22 @@
+####
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/docker/Dockerfile.jvm]: file not necessary to compile

File: quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/docker/Dockerfile.native [hunk 7]
```diff
@@ -0,0 +1,22 @@
+####
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/docker/Dockerfile.native]: file not necessary to compile

File: quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/fruits.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
+function refresh() {
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/fruits.js]: file not necessary to compile

File: quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/streaming.html [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<!doctype html>
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/streaming.html]: file not necessary to compile

File: quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/streaming.js [hunk 10]
```diff
@@ -0,0 +1,11 @@
+if (!!window.EventSource) {
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/streaming.js]: file not necessary to compile

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusProjectInfoCache.java [hunk 11]
```diff
@@ -23,6 +23,7 @@
 import com.redhat.quarkus.commons.QuarkusProjectInfoParams;
 import com.redhat.quarkus.commons.QuarkusPropertiesChangeEvent;
 import com.redhat.quarkus.commons.QuarkusPropertiesScope;
+import com.redhat.quarkus.ls.api.QuarkusProjectInfoProvider;
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusProjectInfoCache.java]: unnecessary import

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusDefinition.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+							List<LocationLink> locations = new ArrayList<>();
+							LocationLink link = new LocationLink(target.getUri(), target.getRange(), target.getRange(),
+									PositionUtils.createRange(key));
+							locations.add(link);
+							return Either.forRight(locations);
+						}
+						// Use simple location
+						List<Location> locations = new ArrayList<>();
+						locations.add(target);
+						return Either.forLeft(locations);
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusDefinition.java]: return Either.forLeft(Collections.singletonList(target))

File: quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesDefinitionTest.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+		value = " #|";
+		testDefinitionFor(value);
+	}
+
+	@Test
+	public void definitionOnKey() throws BadLocationException, InterruptedException, ExecutionException {
+		String value = "quarkus.datasource.driv|er";
+		testDefinitionFor(value, ll(
+				"jdt://contents/quarkus-agroal-0.21.1.jar/io.quarkus.agroal.runtime/DataSourceBuildTimeConfig.class?=all-quarkus-extensions/C:%5C/Users%5C/azerr%5C/.m2%5C/repository%5C/io%5C/quarkus%5C/quarkus-agroal%5C/0.21.1%5C/quarkus-agroal-0.21.1.jar%3Cio.quarkus.agroal.runtime(DataSourceBuildTimeConfig.class",
```
Comment (fbricon) [inline quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesDefinitionTest.java]: how is that hard coded link to your machine supposed to work?

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java (hunk 1 above)
Comment (angelozerr) [reply quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java]: @fbricon updated

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java (hunk 2 above)
Comment (angelozerr) [reply quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java]: @fbricon updated

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/QuarkusDeploymentJavaProject.java (hunk 3 above)
Comment (angelozerr) [reply quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/QuarkusDeploymentJavaProject.java]: @fbricon updated

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/ls/QuarkusDelegateCommandHandler.java (hunk 4 above)
Comment (angelozerr) [reply quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/ls/QuarkusDelegateCommandHandler.java]: @fbricon updated

File: quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/streaming.html (hunk 9 above)
Comment (angelozerr) [reply quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/streaming.html]: @fbricon updated

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusProjectInfoCache.java (hunk 11 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusProjectInfoCache.java]: @fbricon updated

File: quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesDefinitionTest.java (hunk 13 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesDefinitionTest.java]: @fbricon updated

File: quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/streaming.js (hunk 10 above)
Comment (angelozerr) [reply quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/streaming.js]: @fbricon updated

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusDefinition.java (hunk 12 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusDefinition.java]: @fbricon updated

File: quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/fruits.html (hunk 5 above)
Comment (angelozerr) [reply quarkus.jdt/com.redhat.quarkus.jdt.test/projects/maven/using-vertx/src/main/resources/META-INF/resources/fruits.html]: @fbricon updated

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+	private static final Logger LOGGER = Logger.getLogger(MavenArtifactResolver.class.getName());
+
+	@Override
+	public String getArtifact(String groupId, String artifactId, String version) {
+		File jarFile = null;
+		try {
+			jarFile = DependencyUtil.getArtifact(groupId, artifactId, version, null);
+		} catch (Exception e) {
+			LOGGER.log(Level.SEVERE, "Maven artifact JAR download failed.", e);
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java]: maven coordinates should be logged for making troubleshooting easier

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+		return jarFile != null ? jarFile.toString() : null;
+	}
+
+	@Override
+	public String getSources(String groupId, String artifactId, String version) {
+		File jarFile = null;
+		try {
+			jarFile = DependencyUtil.getSources(groupId, artifactId, version);
+		} catch (Exception e) {
+			LOGGER.log(Level.SEVERE, "Maven artifact JAR (Sources) download failed.", e);
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java]: maven coordinates should be logged for making troubleshooting easier

Comment (angelozerr) [reply quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java]: @fbricon updated

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java (hunk 14 above)
Comment (angelozerr) [reply quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/internal/core/MavenArtifactResolver.java]: @fbricon updated

**Comment by angelozerr** [comment]
Thanks @fbricon for your review!

**Pull request merged**